lib/matrix.js

```diff
@@ -52,11 +58,351 @@ var init = function(exports){
+                if (this.rooms[roomId] === undefined) {
+                    this.rooms[roomId] = {};
+                }                
+                if (this.rooms[roomId].timeline === undefined) {
+                    this.rooms[roomId].timeline = [];
+                }
+                if (event.state_key !== undefined) {
+                    this.setStateEvents([events[i]]);
+                }
+                this.rooms[roomId].timeline.push(events[i]);
+            }
+        },
+        
+        /*
+         * Get the timeline of events for a given room
+         * TODO: ordering!
+         */
+        getEvents: function(roomId) {
+            return this.room[roomId].timeline;
+        },
+        
+        setPresenceEvents: function(presenceEvents) {
+            for (var i = 0; i < presenceEvents.length; i++) {
+                var matrixEvent = presenceEvents[i];
+                this.presence[matrixEvent.event.user_id] = matrixEvent;
+            }
+        },
+        
+        getPresenceEvents: function(userId) {
+            return this.presence[userId];
+        },
+        
+        getRoomList: function() {
+            var roomIds = [];
+            for (var roomId in this.rooms) {
+                if (this.rooms.hasOwnProperty(roomId)) {
+                    roomIds.push(roomId);
+                }
+            }
+            return roomIds;
+        },
+        
+        // TODO
+        setMaxHistoryPerRoom: function(maxHistory) {},
+        
+        // TODO
+        reapOldMessages: function() {},
+    };
 
     MatrixClient.prototype = {
         isLoggedIn: function() {
             return this.credentials.accessToken != undefined &&
-            this.credentials.userId != undefined;
+                   this.credentials.userId != undefined;
+        },
+        
+        // Higher level APIs
+        // =================
+        
+        // TODO: stuff to handle:
+        //   local echo
+        //   event dup suppression? - apparently we should still be doing this
+        //   tracking current display name / avatar per-message
+        //   pagination
+        //   re-sending (including persisting pending messages to be sent)
+        //   - Need a nice way to callback the app for arbitrary events like displayname changes
+        //   due to ambiguity (or should this be on a chat-specific layer)?
+                        
+        /*
+         * Helper method for retrieving the name of a room suitable for display in the UI
+         * TODO: in future, this should be being generated serverside.
+         */
+        getFriendlyRoomName: function(roomId) {
+            // we need a store to track the inputs for calculating room names
+            if (!this.store) return roomId;
+            
+            // check for an alias, if any. for now, assume first alias is the official one.
+            var alias;
+            var mRoomAliases = this.store.getStateEvent(roomId, 'm.room.aliases');
```

Comment:
This feels like a bad design choice to make. Can you justify why returning the first matching event in hash order (not even chronologically!!)  is desirable? To me, this seems extremely error prone and makes no sense.